File: src/llama_model_loader.cpp
#include "llama_model_loader.h"
#include "llama_model.h"
#include <godot_cpp/core/class_db.hpp>
#include <godot_cpp/classes/file_access.hpp>
#include <godot_cpp/classes/project_settings.hpp>

using namespace godot;

PackedStringArray LlamaModelLoader::_get_recognized_extensions() const {
	PackedStringArray arr;
	arr.push_back("gguf");
	return arr;
}

Variant godot::LlamaModelLoader::_load(const String &path, const String &original_path, bool use_sub_threads, int32_t cache_mode) const {
	Ref<LlamaModel> model = memnew(LlamaModel);

	if (!FileAccess::file_exists(path)) {
		return ERR_FILE_NOT_FOUND;
	}

	String absPath = ProjectSettings::get_singleton()->globalize_path(path);

	model->load_model(absPath);
	
	return { model };
}

bool LlamaModelLoader::_handles_type(const StringName &type) const {
  return ClassDB::is_parent_class(type, "LlamaModel");
}

String LlamaModelLoader::_get_resource_type(const String &p_path) const {
  String el = p_path.get_extension().to_lower();

  if (el == "gguf") {
    return "LlamaModel";
  }

  return "";
}

File: src/LlamaModelLoader.zig
const std = @import("std");
const Godot = @import("godot");
const Self = @This();

pub usingnamespace Godot.ResourceFormatLoader;
godot_object: *Godot.ResourceFormatLoader,

pub fn _get_recognized_extensions(_: *Self) Godot.PackedStringArray {
    var res = Godot.PackedStringArray.init();
    _ = res.push_back("gguf");
    return res;
}

pub fn _load(_: *Self, path: Godot.String, _: Godot.String, _: bool, _: i32) Godot.Variant {
    const LlamaModel = @import("LlamaModel.zig");
    const model = Godot.create(LlamaModel) catch unreachable;
    // if (Godot.Engine.getSingleton().is_editor_hint()) {
    //     return Godot.Variant.initFrom(.{ .godot_object = model.godot_object });
    // }
    model.load_model(path);
    return Godot.Variant.initFrom(model.*);
}

pub fn _handles_type(_: *Self, type_name: Godot.StringName) bool {
    return type_name.casecmp_to("LlamaModel") == 0;
}

pub fn _get_resource_type(self: *Self, p_path: Godot.String) Godot.String {
    _ = self;
    const el = p_path.get_extension().to_lower();

    if (el.casecmp_to("gguf") == 0) {
        return Godot.String.initFromLatin1Chars("LlamaModel");
    }

    return Godot.String.init();
}
